### ISSUE TITLE
bug:  UnboundLocalError in the dispatcher

### ISSUE THREAD

**Issue by jianshen92** [bentoml/BentoML#4128]
### Describe the bug

```
2023-08-16T18:02:15+0800 [ERROR] [runner:baichuan_runner:1] Traceback (most recent call last):
  File "/root/miniconda3/envs/chen/lib/python3.10/site-packages/bentoml/_internal/marshal/dispatcher.py", line 221, in train_optimizer
    if n < batch_size and (batch_size * a + b) + w0 <= wait:
UnboundLocalError: local variable 'wait' referenced before assignment
```

![Image](https://github.com/bentoml/BentoML/assets/28883559/e42b1af3-6459-451b-846c-96b7eaa80cb1)

This happens sporadically but the application will still run.

### Environment

-

**Comment by KimSoungRyoul** [comment]
this issue can be reproduced when bento spawn runner process more than one

like a 

~~~yaml
runners:
  resources:
    cpu: 4
  workers_per_resource: 2 # more than one

~~~

or 

~~~yaml
runners:
  resources:
    nvidia.com/gpu: 2
  workers_per_resource: 2 # more than one
~~~

**Comment by Matthieu-Tinycoaching** [comment]
```
Ubuntu: 22.04.03 LTS
Python: 3.11.5
bentoML: 1.1.6
```

Hi,

I got the same error message:
```
2023-09-12T15:27:01+0200 [INFO] [cli] Environ for worker 0: set CUDA_VISIBLE_DEVICES to 0
2023-09-12T15:27:01+0200 [INFO] [cli] Environ for worker 0: set CUDA_VISIBLE_DEVICES to 0
2023-09-12T15:27:01+0200 [INFO] [cli] Prometheus metrics for HTTP BentoServer from "bentoml_gpu_ct2_mlc_vicuna-7b-v1-5_service.py:ct2_mlc_vic7bs" can be accessed at http://localhost:3000/metrics.
2023-09-12T15:27:03+0200 [INFO] [cli] Starting production HTTP BentoServer from "bentoml_gpu_ct2_mlc_vicuna-7b-v1-5_service.py:ct2_mlc_vic7bs" listening on http://0.0.0.0:3000 (Press CTRL+C to quit)
2023-09-12T15:28:37+0200 [INFO] [runner:ct2_vicuna7b_int8_float16_scores_bt_runner:1] _ (scheme=http,method=POST,path=/,type=application/octet-stream,length=2004) (status=200,type=application/vnd.bentoml.multiple_outputs,length=1331) 2826.426ms (trace=fcf4f0493039c3cb3a54e35a173b9786,span=551022755e8864bd,sampled=0,service.name=ct2_vicuna7b_int8_float16_scores_bt_runner)
2023-09-12T15:28:37+0200 [INFO] [api_server:ct2_mlc_vicuna7b_service:36] 127.0.0.1:59472 (scheme=http,method=POST,path=/generate_ct2_vic7b_int8_float16_scores_bt_async,type=text/plain; charset=utf-8,length=2259) (status=200,type=application/json,length=1247) 2886.239ms (trace=fcf4f0493039c3cb3a54e35a173b9786,span=08e7924917097c0a,sampled=0,service.name=ct2_mlc_vicuna7b_service)
2023-09-12T15:28:41+0200 [ERROR] [runner:ct2_vicuna7b_int8_float16_scores_bt_runner:1] Traceback (most recent call last):
  File "/home/matthieu/anaconda3/envs/bentoml_1.1.6_gpu/lib/python3.11/site-packages/bentoml/_internal/marshal/dispatcher.py", line 221, in train_optimizer
    if n < batch_size and (batch_size * a + b) + w0 <= wait:
                                                       ^^^^
UnboundLocalError: cannot access local variable 'wait' where it is not associated with a value
 (trace=fcf4f0493039c3cb3a54e35a173b9786,span=551022755e8864bd,sampled=0,service.name=ct2_vicuna7b_int8_float16_scores_bt_runner)
Traceback (most recent call last):
  File "/home/matthieu/anaconda3/envs/bentoml_1.1.6_gpu/lib/python3.11/site-packages/bentoml/_internal/marshal/dispatcher.py", line 221, in train_optimizer
    if n < batch_size and (batch_size * a + b) + w0 <= wait:
                                                       ^^^^
UnboundLocalError: cannot access local variable 'wait' where it is not associated with a value
2023-09-12T15:28:44+0200 [INFO] [runner:ct2_vicuna7b_int8_float16_scores_bt_runner:1] _ (scheme=http,method=POST,path=/,type=application/octet-stream,length=2004) (status=200,type=application/vnd.bentoml.multiple_outputs,length=1331) 2747.621ms (trace=8ef7182c54a86f1862fb46edd5830712,span=98edfdc48058b9f2,sampled=0,service.name=ct2_vicuna7b_int8_float16_scores_bt_runner)
2023-09-12T15:28:44+0200 [INFO] [api_server:ct2_mlc_vicuna7b_service:47] 127.0.0.1:59478 (scheme=http,method=POST,path=/generate_ct2_vic7b_int8_float16_scores_bt_async,type=text/plain; charset=utf-8,length=2259) (status=200,type=application/json,length=1247) 2779.012ms (trace=8ef7182c54a86f1862fb46edd5830712,span=fd9eb863af688f8a,sampled=0,service.name=ct2_mlc_vicuna7b_service)
```

When running: `bentoml serve bentoml_gpu_ct2_mlc_vicuna-7b-v1-5_service.py:ct2_mlc_vic7bs --reload`

**Comment by Hubert-Bonisseur** [comment]
I have the same issue with this very simple service made to test adaptative batching with custom runners:

service.py
```python
import time
from ctypes import Array

import bentoml
from bentoml.io import NumpyNdarray


class MyRunnable(bentoml.Runnable):
    SUPPORTED_RESOURCES = ("nvidia.com/gpu", "cpu")
    SUPPORTS_CPU_MULTI_THREADING = True

    @bentoml.Runnable.method(batchable=True)
    def identity(self, batch: Array) -> Array:
        print("batch", batch)
        time.sleep(5)
        return batch


runner = bentoml.Runner(MyRunnable)
svc = bentoml.Service("testing", runners=[runner])


@svc.api(input=NumpyNdarray(), output=NumpyNdarray())
async def predict(input):
    print("input", input)
    prediction = await runner.identity.async_run(input)
    print("raw_predict", prediction)
    return prediction
```

call api in quick succession with this code

```python
import requests

url = "http://127.0.0.1:3000/predict"
headers = {"content-type": "application/json"}
data = [[1, 1, 1], [2, 2, 2], [3, 3, 3], [4, 4, 4], [5, 5, 5]]

for i in range(len(data)):
    response = requests.post(url, headers=headers, json=data[i])
    print("Response:", response.text)
```

I have this config:
configuration.yaml
```
runners:
  batching:
    enabled: true
    max_batch_size: 10
    max_latency_ms: 10000
```

And I also get the same error message:

```
  File "[redacted]/LanguageDetector/venv/lib/python3.9/site-packages/bentoml/_internal/marshal/dispatcher.py", line 221, in train_optimizer
    if n < batch_size and (batch_size * a + b) + w0 <= wait:
UnboundLocalError: local variable 'wait' referenced before assignment
```

**Comment by northeastloon** [comment]
I also get this error when sending concurrent requests to the service. My service.py is defined as follows:

```
import bentoml

from bentoml.io import Text, JSON

runner = bentoml.transformers.get("classifier:latest").to_runner()

svc = bentoml.Service("classifier_service", runners=[runner])

@svc.api(input=JSON(), output=JSON())
async def classify(input_json: list) -> list:
    results = await runner.async_run(input_json)
    return results
```

**Comment by aarnphm** [comment]
This is now fixed and released on 1.1.10
